### PR TITLE
Issue 6: Allow setting of digest and label for OAEP encryption and add CHACHA20 unwrap

### PR DESCRIPTION
**Pull request by ericeberry**

### CODE REVIEW

File: reference/src/client/src/sa_engine_pkey.c [hunk 1]
```diff
@@ -137,13 +153,32 @@ static int pkey_copy(
     new_app_data->pss_salt_length = app_data->pss_salt_length;
     new_app_data->evp_md_ctx = app_data->evp_md_ctx;
     new_app_data->evp_md = app_data->evp_md;
+    new_app_data->oaep_md = app_data->oaep_md;
+    new_app_data->oaep_mgf1_md = app_data->oaep_mgf1_md;
+    if (app_data->oaep_label_length > 0) {
+        new_app_data->oaep_label = OPENSSL_malloc(app_data->oaep_label_length);
+        if (new_app_data->oaep_label == NULL) {
+            ERROR("malloc failed");
```
Comment (mhabrat) [inline reference/src/client/src/sa_engine_pkey.c:161]: OPENSSL_malloc

File: reference/src/client/test/client_test_helpers.h [hunk 2]
```diff
@@ -400,12 +400,19 @@ namespace client_test_helpers {
      * @param[out] out output buffer.
      * @param[in] in input buffer.
      * @param[in] evp_pkey RSA key.
+     * @param[in] digest_algorithm the digest algorithm for OAEP padding.
+     * @param[in] mgf1_digest_algorithm the digest algorithm for the MGF1 function.
+     * @param[in] label the label for the OAEP padding. May be NULL.
+     * @param[in] label_length the length of the label. Must be 0 if label is NULL.
```
Comment (mhabrat) [inline reference/src/client/test/client_test_helpers.h:406]: Is this extraneous or needed in the function prototype?

File: reference/src/client/test/sa_key_unwrap_chacha20.cpp [hunk 3]
```diff
@@ -0,0 +1,364 @@
+/**
+ * Copyright 2020-2021 Comcast Cable Communications Management, LLC
```
Comment (mhabrat) [inline reference/src/client/test/sa_key_unwrap_chacha20.cpp:2]: Extend copyright to 2022.

File: reference/src/client/test/sa_key_unwrap_chacha20_poly1305.cpp [hunk 4]
```diff
@@ -0,0 +1,417 @@
+/**
+ * Copyright 2020-2021 Comcast Cable Communications Management, LLC
```
Comment (mhabrat) [inline reference/src/client/test/sa_key_unwrap_chacha20_poly1305.cpp:2]: Extend to 2022.

File: reference/src/taimpl/src/internal/unwrap.c [hunk 5]
```diff
@@ -473,13 +473,278 @@ sa_status unwrap_aes_gcm(
...
+        context = EVP_CIPHER_CTX_new();
+        if (context == NULL) {
+            ERROR("EVP_CIPHER_CTX_new failed");
+            status = SA_STATUS_INTERNAL_ERROR;
+            break;
+        }
+
+        const EVP_CIPHER* cipher = EVP_chacha20_poly1305();
+        if (cipher == NULL) {
+            ERROR("EVP_chacha20 failed");
```
Comment (mhabrat) [inline reference/src/taimpl/src/internal/unwrap.c:651]: EVP_chacha20_poly1305

File: reference/src/taimpl/src/internal/unwrap.c [hunk 6]
```diff
@@ -473,13 +473,278 @@ sa_status unwrap_aes_gcm(
...
+
+        if (EVP_DecryptInit_ex(context, cipher, NULL, NULL, NULL) != 1) {
+            ERROR("EVP_DecryptInit_ex failed");
+            status = SA_STATUS_INTERNAL_ERROR;
+            break;
+        }
+
+        // set nonce length
+        if (EVP_CIPHER_CTX_ctrl(context, EVP_CTRL_AEAD_SET_IVLEN, (int) parameters->nonce_length, NULL) != 1) {
+            ERROR("EVP_CIPHER_CTX_counterl failed");
```
Comment (mhabrat) [inline reference/src/taimpl/src/internal/unwrap.c:664]: EVP_CIPHER_CTX_ctrl

File: reference/src/taimpl/src/internal/unwrap.c [hunk 7]
```diff
@@ -473,13 +473,278 @@ sa_status unwrap_aes_gcm(
...
+            status = SA_STATUS_INTERNAL_ERROR;
+            break;
+        }
+
+        // init key and nonce
+        if (EVP_DecryptInit_ex(context, cipher, NULL, key, parameters->nonce) != 1) {
+            ERROR("EVP_EncryptInit_ex failed");
```
Comment (mhabrat) [inline reference/src/taimpl/src/internal/unwrap.c:671]: EVP_DecryptInit_ex

File: reference/src/taimpl/src/internal/unwrap.c [hunk 8]
```diff
@@ -473,13 +473,278 @@ sa_status unwrap_aes_gcm(
...
+            ERROR("EVP_CIPHER_CTX_set_padding failed");
+            status = SA_STATUS_INTERNAL_ERROR;
+            break;
+        }
+
+        // set aad
+        if (parameters->aad != NULL) {
+            int out_length = 0;
+            if (EVP_DecryptUpdate(context, NULL, &out_length, parameters->aad, (int) parameters->aad_length) != 1) {
+                ERROR("EVP_EncryptUpdate failed");
```
Comment (mhabrat) [inline reference/src/taimpl/src/internal/unwrap.c:687]: EVP_DecryptUpdate

File: reference/src/taimpl/src/ta_sa_crypto_cipher_init.c [hunk 9]
```diff
@@ -694,28 +693,115 @@ static sa_status ta_sa_crypto_cipher_init_rsa(
...
+        status = cipher_store_add_asymmetric_key(context, cipher_store, SA_CIPHER_ALGORITHM_RSA_OAEP, cipher_mode,
+                stored_key, caller_uuid);
         if (status != SA_STATUS_OK) {
             ERROR("cipher_store_add_asymmetric_key failed");
             break;
         }
+
+        status = cipher_store_acquire_exclusive(&cipher, cipher_store, *context, caller_uuid);
+        if (status != SA_STATUS_OK) {
+            ERROR("cipher_store_add_asymmetric_key failed");
```
Comment (mhabrat) [inline reference/src/taimpl/src/ta_sa_crypto_cipher_init.c:790]: cipher_store_acquire_exclusive

File: reference/src/taimpl/src/ta_sa_crypto_cipher_init.c [hunk 10]
```diff
@@ -905,9 +991,15 @@ sa_status ta_sa_crypto_cipher_init(
...
+            if (status != SA_STATUS_OK) {
+                ERROR("ta_sa_crypto_cipher_init_rsa failed");
+                break;
+            }
+        } else if (cipher_algorithm == SA_CIPHER_ALGORITHM_RSA_OAEP) {
+            status = ta_sa_crypto_cipher_init_rsa_oaep(context, cipher_mode, stored_key,
+                    (sa_cipher_parameters_rsa_oaep*) parameters, client,
                     caller_uuid);
             if (status != SA_STATUS_OK) {
                 ERROR("ta_sa_crypto_cipher_init_rsa failed");
```
Comment (mhabrat) [inline reference/src/taimpl/src/ta_sa_crypto_cipher_init.c:1005]: ta_sa_crypto_cipher_init_rsa_oaep

File: reference/src/taimpl/src/ta_sa_key_unwrap.c [hunk 11]
```diff
@@ -462,11 +462,246 @@ static sa_status ta_sa_key_unwrap_aes_gcm(
...
+        if (!key_type_supports_chacha20(header->type, header->size)) {
+            ERROR("key_type_supports_chacha20 failed");
+            status = SA_STATUS_BAD_KEY_TYPE;
+            break;
+        }
+
+        status = unwrap_chacha20(&stored_key_unwrapped, in, in_length, rights, key_type, type_parameters,
+                algorithm_parameters, stored_key_wrapping);
+        if (status != SA_STATUS_OK) {
+            ERROR("unwrap_aes_ctr failed");
```
Comment (mhabrat) [inline reference/src/taimpl/src/ta_sa_key_unwrap.c:563]: unwrap_chacha20(

File: reference/src/taimpl/src/ta_sa_key_unwrap.c [hunk 12]
```diff
@@ -462,11 +462,246 @@ static sa_status ta_sa_key_unwrap_aes_gcm(
...
+        if (!key_type_supports_chacha20(header->type, header->size)) {
+            ERROR("key_type_supports_chacha20 failed");
+            status = SA_STATUS_BAD_KEY_TYPE;
+            break;
+        }
+
+        status = unwrap_chacha20_poly1305(&stored_key_unwrapped, in, in_length, rights, key_type, type_parameters,
+                algorithm_parameters, stored_key_wrapping);
+        if (status != SA_STATUS_OK) {
+            ERROR("unwrap_aes_gcm failed");
```
Comment (mhabrat) [inline reference/src/taimpl/src/ta_sa_key_unwrap.c:681]: unwrap_chacha20_poly1305

File: reference/src/taimpl/src/ta_sa_key_unwrap.c [hunk 13]
```diff
@@ -739,10 +974,28 @@ sa_status ta_sa_key_unwrap(
                 ERROR("ta_sa_key_unwrap_aes_gcm failed");
                 break;
             }
+#if OPENSSL_VERSION_NUMBER >= 0x10100000
+        } else if (cipher_algorithm == SA_CIPHER_ALGORITHM_CHACHA20) {
+            status = ta_sa_key_unwrap_chacha20(key, rights, key_type, type_parameters,
+                    (sa_unwrap_parameters_chacha20*) algorithm_parameters, wrapping_key, in,
+                    in_length, client, caller_uuid);
+            if (status != SA_STATUS_OK) {
+                ERROR("ta_sa_key_unwrap_aes_ctr failed");
```
Comment (mhabrat) [inline reference/src/taimpl/src/ta_sa_key_unwrap.c:983]: ta_sa_key_unwrap_chacha20

File: reference/src/taimpl/src/ta_sa_key_unwrap.c [hunk 14]
```diff
@@ -739,10 +974,28 @@ sa_status ta_sa_key_unwrap(
...
+                break;
+            }
+        } else if (cipher_algorithm == SA_CIPHER_ALGORITHM_CHACHA20_POLY1305) {
+            status = ta_sa_key_unwrap_chacha20_poly1305(key, rights, key_type, type_parameters,
+                    (sa_unwrap_parameters_chacha20_poly1305*) algorithm_parameters, wrapping_key, in,
+                    in_length, client, caller_uuid);
+            if (status != SA_STATUS_OK) {
+                ERROR("ta_sa_key_unwrap_aes_gcm failed");
```
Comment (mhabrat) [inline reference/src/taimpl/src/ta_sa_key_unwrap.c:991]: ta_sa_key_unwrap_chacha20_poly1305

**Review by mhabrat — CHANGES_REQUESTED** [review]

File: reference/src/client/src/sa_engine_pkey.c (hunk 1 above)
Comment (ericeberry) [reply reference/src/client/src/sa_engine_pkey.c:161]: Done

File: reference/src/client/test/client_test_helpers.h (hunk 2 above)
Comment (ericeberry) [reply reference/src/client/test/client_test_helpers.h:406]: Extraneous. Done

File: reference/src/client/test/sa_key_unwrap_chacha20.cpp (hunk 3 above)
Comment (ericeberry) [reply reference/src/client/test/sa_key_unwrap_chacha20.cpp:2]: Done

File: reference/src/client/test/sa_key_unwrap_chacha20_poly1305.cpp (hunk 4 above)
Comment (ericeberry) [reply reference/src/client/test/sa_key_unwrap_chacha20_poly1305.cpp:2]: Done

File: reference/src/taimpl/src/internal/unwrap.c (hunk 5 above)
Comment (ericeberry) [reply reference/src/taimpl/src/internal/unwrap.c:651]: Done

File: reference/src/taimpl/src/internal/unwrap.c (hunk 6 above)
Comment (ericeberry) [reply reference/src/taimpl/src/internal/unwrap.c:664]: Done

File: reference/src/taimpl/src/internal/unwrap.c (hunk 7 above)
Comment (ericeberry) [reply reference/src/taimpl/src/internal/unwrap.c:671]: Done

File: reference/src/taimpl/src/internal/unwrap.c (hunk 8 above)
Comment (ericeberry) [reply reference/src/taimpl/src/internal/unwrap.c:687]: Done

File: reference/src/taimpl/src/ta_sa_crypto_cipher_init.c (hunk 10 above)
Comment (ericeberry) [reply reference/src/taimpl/src/ta_sa_crypto_cipher_init.c:1005]: Done

File: reference/src/taimpl/src/ta_sa_key_unwrap.c (hunk 11 above)
Comment (ericeberry) [reply reference/src/taimpl/src/ta_sa_key_unwrap.c:563]: Done

File: reference/src/taimpl/src/ta_sa_key_unwrap.c (hunk 12 above)
Comment (ericeberry) [reply reference/src/taimpl/src/ta_sa_key_unwrap.c:681]: Done

File: reference/src/taimpl/src/ta_sa_key_unwrap.c (hunk 13 above)
Comment (ericeberry) [reply reference/src/taimpl/src/ta_sa_key_unwrap.c:983]: Done

File: reference/src/taimpl/src/ta_sa_key_unwrap.c (hunk 14 above)
Comment (ericeberry) [reply reference/src/taimpl/src/ta_sa_key_unwrap.c:991]: Done

File: reference/src/taimpl/src/ta_sa_crypto_cipher_init.c (hunk 9 above)
Comment (ericeberry) [reply reference/src/taimpl/src/ta_sa_crypto_cipher_init.c:790]: Done

**Review by mhabrat — APPROVED** [review]

**Pull request merged**